sdk/deviceprovisioning/azure-iothub-deviceprovisioning/src/main/java/com/azure/deviceprovisioning/DeviceProvisioningServiceClientBuilder.java

```diff
@@ -0,0 +1,364 @@
+
+    private Configuration configuration;
+
+    /**
+     * The public constructor for DeviceProvisioningServiceClientBuilder
+     */
+    public DeviceProvisioningServiceClientBuilder()
+    {
+        additionalPolicies = new ArrayList<>();
+        properties = CoreUtils.getProperties(DEVICE_PROVISIONING_SERVICE_PROPERTIES);
+        httpLogOptions = new HttpLogOptions();
+    }
+
+    private static HttpPipeline buildPipeline(AzureSasCredential azureSasCredential,
+                                              String sasToken,
+                                              HttpLogOptions httpLogOptions,
+                                              ClientOptions clientOptions,
+                                              HttpClient httpClient,
+                                              List<HttpPipelinePolicy> additionalPolicies,
+                                              RetryPolicy retryPolicy,
+                                              Configuration configuration,
+                                              Map<String, String> properties) {
+        // Closest to API goes first, closest to wire goes last.
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+
+        String clientName = properties.getOrDefault(SDK_NAME, "UnknownName");
+        String clientVersion = properties.getOrDefault(SDK_VERSION, "UnknownVersion");
+
+        // Give precedence to applicationId configured in clientOptions over the one configured in httpLogOptions.
+        // Azure.Core deprecated setting the applicationId in httpLogOptions, but we should still support it.
+        String applicationId = clientOptions == null
+            ? httpLogOptions.getApplicationId()
+            : clientOptions.getApplicationId();
+
+        policies.add(new UserAgentPolicy(applicationId, clientName, clientVersion, configuration));
+
+        // Adds a "x-ms-client-request-id" header to each request. This header is useful for tracing requests through Azure ecosystems
+        policies.add(new RequestIdPolicy());
+
+        // Only the RequestIdPolicy  and UserAgentPolicy will take effect prior to the retry policy since neither of those need
+        // to change in any way upon retry
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+        policies.add(retryPolicy);
+
+        // Adds a date header to each HTTP request for tracking purposes
+        policies.add(new AddDatePolicy());
+
+        // Add authentication policy so that each HTTP request has authorization header
+        HttpPipelinePolicy credentialPolicy;
+        if (azureSasCredential != null) {
+            credentialPolicy = new AzureSasCredentialPolicy(azureSasCredential, false);
+        } else if (sasToken != null) {
+            credentialPolicy = new AzureSasCredentialPolicy(new AzureSasCredential(sasToken), false);
+        } else {
+            credentialPolicy =  null;
+        }
+
+        if (credentialPolicy != null) {
```

Comment:
It seems odd to not make this mandatory, but I saw this pattern in the Storage SDK client builders. Any thoughts, @srnagar ?

Reply:
This should be required. If credentialPolicy is null, then a valid client cannot be constructed and the build method should throw. I am not sure why storage is doing this but I'll follow up with them.